File: openmrs-client/src/main/java/org/openmrs/mobile/models/Person.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */

package org.openmrs.mobile.models;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import android.graphics.Bitmap;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import org.openmrs.mobile.utilities.ImageUtils;

public class Person extends Resource implements Serializable {

    @SerializedName("names")
    @Expose
    private List<PersonName> names = new ArrayList<>();
    @SerializedName("gender")
    @Expose
    private String gender;
    @SerializedName("birthdate")
    @Expose
    private String birthdate;
    @SerializedName("birthdateEstimated")
    @Expose
    private boolean birthdateEstimated;
    @SerializedName("addresses")
    @Expose
    private List<PersonAddress> addresses = new ArrayList<>();
    @SerializedName("attributes")
    @Expose
    private List<PersonAttribute> attributes = new ArrayList<>();

    private Bitmap photo;

    public Person() {
    }

    public Person(List<PersonName> names, String gender, String birthdate, boolean birthdateEstimated, List<PersonAddress> addresses, List<PersonAttribute> attributes, Bitmap photo) {
        this.names = names;
        this.gender = gender;
        this.birthdate = birthdate;
        this.birthdateEstimated = birthdateEstimated;
        this.addresses = addresses;
        this.attributes = attributes;
        this.photo = photo;
    }

    /**
     * @return The names
     */
    public List<PersonName> getNames() {
        return names;
    }

    /**
     * @param names The names
     */
    public void setNames(List<PersonName> names) {
        this.names = names;
    }

    public PersonName getName() {
        if (!names.isEmpty()) {
            return names.get(0);
        } else {
            return null;
        }
    }

    /**
     * @return The gender
     */
    public String getGender() {
        return gender;
    }

    /**
     * @param gender The gender
     */
    public void setGender(String gender) {
        this.gender = gender;
    }

    /**
     * @return The birthdate
     */
    public String getBirthdate() {
        return birthdate;
    }

    /**
     * @param birthdate The birthdate
     */
    public void setBirthdate(String birthdate) {
        this.birthdate = birthdate;
    }

    /**
     * @return The birthdateEstimated
     */
    public boolean getBirthdateEstimated() {
        return birthdateEstimated;
    }

    /**
     * @param birthdateEstimated The birthdate
     */
    public void setBirthdateEstimated(boolean birthdateEstimated) {
        this.birthdateEstimated = birthdateEstimated;
    }


    /**
     * @return The addresses
     */
    public List<PersonAddress> getAddresses() {
        return addresses;
    }

    /**
     * @param addresses The addresses
     */
    public void setAddresses(List<PersonAddress> addresses) {
        this.addresses = addresses;
    }

    public PersonAddress getAddress() {
        if (!addresses.isEmpty()) {
            return addresses.get(0);
        } else {
            return null;
        }
    }

    /**
     * @return The attributes
     */
    public List<PersonAttribute> getAttributes() {
        return attributes;
    }

    /**
     * @param attributes The attributes
     */
    public void setAttributes(List<PersonAttribute> attributes) {
        this.attributes = attributes;
    }


    public Bitmap getPhoto() {
        return photo;
    }

    public void setPhoto(Bitmap patientPhoto) {
        this.photo = patientPhoto;
    }

    public Bitmap getResizedPhoto() {
        return ImageUtils.resizePhoto(this.photo);
    }
}

File: openmrs-client/src/main/java/org/openmrs/mobile/models/Person.kt
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */
package org.openmrs.mobile.models

import android.graphics.Bitmap
import androidx.room.TypeConverters
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import org.openmrs.mobile.models.typeConverters.PersonAddressConverter
import org.openmrs.mobile.models.typeConverters.PersonAttributeConverter
import org.openmrs.mobile.models.typeConverters.PersonNameConverter
import org.openmrs.mobile.utilities.ImageUtils.resizePhoto
import java.io.Serializable
import java.util.*

open class Person : Resource, Serializable {
    /**
     * @return The names
     */
    /**
     * @param names The names
     */
    @TypeConverters(PersonNameConverter::class)
    @SerializedName("names")
    @Expose
    var names: MutableList<PersonName> = ArrayList()

    /**
     * @return The gender
     */
    /**
     * @param gender The gender
     */
    @SerializedName("gender")
    @Expose
    var gender: String? = null

    /**
     * @return The birthdate
     */
    /**
     * @param birthdate The birthdate
     */
    @SerializedName("birthdate")
    @Expose
    var birthdate: String? = null

    /**
     * @return The birthdateEstimated
     */
    /**
     * @param birthdateEstimated The birthdate
     */
    @SerializedName("birthdateEstimated")
    @Expose
    var birthdateEstimated = false

    /**
     * @return The addresses
     */
    /**
     * @param addresses The addresses
     */
    @TypeConverters(PersonAddressConverter::class)
    @SerializedName("addresses")
    @Expose
    var addresses: MutableList<PersonAddress> = ArrayList()

    /**
     * @return The attributes
     */
    /**
     * @param attributes The attributes
     */
    @TypeConverters(PersonAttributeConverter::class)
    @SerializedName("attributes")
    @Expose
    var attributes: List<PersonAttribute> = ArrayList()

    @SerializedName("dead")
    @Expose
    var isDeceased: Boolean? = null

    @SerializedName("causeOfDeath")
    @Expose
    var causeOfDeath: Resource? = null
    var photo: Bitmap? = null

    constructor()
    constructor(names: MutableList<PersonName>, gender: String?, birthdate: String?, birthdateEstimated: Boolean, addresses: MutableList<PersonAddress>, attributes: List<PersonAttribute>,
                photo: Bitmap?, causeOfDeath: Resource?, dead: Boolean) {
        this.names = names
        this.gender = gender
        this.birthdate = birthdate
        this.birthdateEstimated = birthdateEstimated
        this.addresses = addresses
        this.attributes = attributes
        this.photo = photo
        this.causeOfDeath = causeOfDeath
        isDeceased = dead
    }

    val name: PersonName?
        get() = if (names.isNotEmpty()) {
            names[0]
        } else {
            null
        }

    val address: PersonAddress?
        get() = if (addresses.isNotEmpty()) {
            addresses[0]
        } else {
            null
        }

    val resizedPhoto: Bitmap
        get() = resizePhoto(photo!!)

}